### PR TITLE
Fixes for issue#173. Improved robustness of message sending

### PR DESCRIPTION
**Pull request by vgtom**

### CODE REVIEW

File: ethcore/sync/src/api.rs [hunk 1]
```diff
@@ -590,13 +590,21 @@ impl ChainNotify for EthSync {
...
+											None => { warn!(target:"sync", "No session exists for peerId {:?}", x); false},
+											Some(n) => true,
+										} 
+									}).collect::<Vec<_>>();
+			let node_ids = filter_peers.iter()
+								.map(|&x|context.session_info(*x).unwrap().id)
+								.collect::<Vec<Option<H512>>>();
 			let index = node_ids.into_iter().position(|r| r == node_id).unwrap();
-			let my_peer_id = peer_ids[index];
+			let my_peer_id = filter_peers[index];
```
Comment (dforsten) [inline ethcore/sync/src/api.rs]: Still there is use of unwrap() which will cause crashes at runtime.

It is a lot of code for finding a peer id from a collection of peer ids, another option would be to just use "find()", for example:
```
			let peer_ids = self.network.connected_peers();
			let target_peer_id = peer_ids.iter().find(|p| {
				match context.session_info(**p){
					Some(session_info) => {
						session_info.id == node_id
					},
					None => { warn!(target:"sync", "No session exists for peerId {:?}", p); false},
				}
			});
```

**Comment by vgtom** [comment]
Could you review and check please.
The unimplemented part i will make a commit today to store it in a cache and send when a connection is established.

**Comment by dforsten** [comment]
Implementing the targeted message cache is a separate issue and pull request, we should get this stability issue sorted out first.

unimplemented! will panic the application, please replace it with something that will not crash. You could issue a warn! and return from the function in that case for example.

**Comment by vgtom** [comment]
Please check now

**Pull request merged**